### PR TITLE
feat: events by slices

### PR DESCRIPTION
**Pull request by patriknw**

### CODE REVIEW

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala [hunk 1]
```diff
@@ -71,4 +79,86 @@ import software.amazon.awssdk.services.dynamodb.model.QueryRequest
...
+                  tags = Set.empty, // FIXME
+                  metadata = metadata)
+              }
+            }
+          }
+        }
+        .mapMaterializedValue(_ => NotUsed)
+    }
+
+    queryOneSlice(minSlice).concatAllLazy((minSlice + 1 to maxSlice).map(queryOneSlice): _*)
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala]: This doesn't work. Each query (per slice) need to keep track of current timestamp offset, and now it's shared in the BySliceQuery.State. I will change to have one BySliceQuery per slice instead, and merge at outer layer.

File: core/src/main/scala/akka/persistence/dynamodb/internal/SerializedJournalItem.scala [hunk 2]
```diff
@@ -30,7 +36,7 @@ final case class SerializedEventMetadata(serId: Int, serManifest: String, payloa
   val SeqNr = "seq_nr"
   // needed for the bySlices GSI
   val EntityTypeSlice = "entity_type_slice"
-  val Timestamp = "timestamp"
+  val Timestamp = "ts"
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/dynamodb/internal/SerializedJournalItem.scala:39]: some names are reserved

**Comment by pvlugter** [comment]
Reminds me how complex the behaviour of the events by slices can be, and that we're not sure whether we resolved all the projection stalling issues. Wonder if the backpressure from GSI to the base table would actually end up helping here as well. Not something for now, but gets me pondering alternative overall designs that could simplify this side of projections.

File: core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala (hunk 1 above)
Comment (patriknw) [reply core/src/main/scala/akka/persistence/dynamodb/internal/QueryDao.scala]: merging at the outer layer and single slice per BySliceQuery worked

**Comment by patriknw** [comment]
This should be ready for final review. I'll follow up with some related things in separate PR.

File: core/src/main/resources/reference.conf [hunk 3]
```diff
@@ -29,6 +29,29 @@ akka.persistence.dynamodb {
...
+
+    # When live queries return no results or <= 10% of buffer-size, the next query
+    # to db will be delayed for this duration.
+    # When the number of rows from previous query is >= 90% of buffer-size, the next
+    # query will be emitted immediately.
+    # Otherwise, between 10% - 90% of buffer-size, the next query will be delayed
+    # for half of this duration.
+    refresh-interval = 3s
+
+    # Live queries read events up to this duration from the current database time.
```
Comment (pvlugter) [inline core/src/main/resources/reference.conf]: Not important, but should we just say `current time` wherever we mention `current database time`?

File: core/src/main/scala/akka/persistence/dynamodb/internal/BySliceQuery.scala [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+import akka.persistence.query.Offset
+import akka.persistence.query.TimestampOffset
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.Source
+import org.slf4j.Logger
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[dynamodb] object BySliceQuery {
```
Comment (pvlugter) [inline core/src/main/scala/akka/persistence/dynamodb/internal/BySliceQuery.scala:27]: Also read this as a diff with the r2dbc version. Looks good.

File: core/src/main/scala/akka/persistence/dynamodb/internal/ContinuousQuery.scala [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+import akka.stream.SourceShape
+import akka.stream.scaladsl.Source
+import akka.stream.stage._
+import akka.util.OptionVal
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[dynamodb] object ContinuousQuery {
```
Comment (pvlugter) [inline core/src/main/scala/akka/persistence/dynamodb/internal/ContinuousQuery.scala:27]: Same as r2dbc.

File: core/src/main/scala/akka/persistence/dynamodb/query/javadsl/DynamodbReadJournal.scala [hunk 6]
```diff
@@ -4,11 +4,75 @@
...
+   * Query events for given slices. A slice is deterministically defined based on the persistence id. The purpose is to
+   * evenly distribute all persistence ids over the slices.
+   *
+   * The consumer can keep track of its current position in the event stream by storing the `offset` and restart the
+   * query from a given `offset` after a crash/restart.
+   *
+   * The supported offset is [[akka.persistence.query.TimestampOffset]] and [[Offset.noOffset]].
+   *
+   * The timestamp is based on the client wall clock and the events are read from a DynamoDB global secondary index,
+   * which is eventually consistent.This means that a "later" event may be visible first and when retrieving events
```
Comment (pvlugter) [inline core/src/main/scala/akka/persistence/dynamodb/query/javadsl/DynamodbReadJournal.scala]:
```suggestion
   * which is eventually consistent. This means that a "later" event may be visible first and when retrieving events
```

File: core/src/main/scala/akka/persistence/dynamodb/query/scaladsl/DynamoDBReadJournal.scala [hunk 7]
```diff
@@ -44,4 +96,66 @@ final class DynamoDBReadJournal(system: ExtendedActorSystem, config: Config, cfg
...
+   * Query events for given slices. A slice is deterministically defined based on the persistence id. The purpose is to
+   * evenly distribute all persistence ids over the slices.
+   *
+   * The consumer can keep track of its current position in the event stream by storing the `offset` and restart the
+   * query from a given `offset` after a crash/restart.
+   *
+   * The supported offset is [[TimestampOffset]] and [[Offset.noOffset]].
+   *
+   * The timestamp is based on the client wall clock and the events are read from a DynamoDB global secondary index,
+   * which is eventually consistent.This means that a "later" event may be visible first and when retrieving events
```
Comment (pvlugter) [inline core/src/main/scala/akka/persistence/dynamodb/query/scaladsl/DynamoDBReadJournal.scala]:
```suggestion
   * which is eventually consistent. This means that a "later" event may be visible first and when retrieving events
```

File: core/src/test/scala/akka/persistence/dynamodb/query/EventsBySliceBacktrackingSpec.scala [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+      env5.persistenceId shouldBe pid1.id
+      env5.sequenceNr shouldBe 3L
+
+      // before e1-3 so it will not be found by the normal query
+      writeEvent(slice, pid2, 1L, startTime.plusMillis(2), "e2-1")
+
+      // no backtracking yet
+      result.expectNoMessage(settings.querySettings.refreshInterval + 100.millis)
+
+      // after 1/2 of the backtracking widow, to kick off a backtracking query
```
Comment (pvlugter) [inline core/src/test/scala/akka/persistence/dynamodb/query/EventsBySliceBacktrackingSpec.scala]: Just tests, but while noticed

```suggestion
      // after 1/2 of the backtracking window, to kick off a backtracking query
```

File: core/src/main/resources/reference.conf (hunk 3 above)
Comment (pvlugter) [reply core/src/main/resources/reference.conf]: Will adjust these while here, since we're only using app time.

Comment (pvlugter) [reply core/src/main/resources/reference.conf]:
```suggestion
    # Live queries read events up to this duration from the current time.
```

File: core/src/main/resources/reference.conf [hunk 9]
```diff
@@ -29,6 +29,29 @@ akka.persistence.dynamodb {
...
+    behind-current-time = 100 millis
+
+    backtracking {
+      enabled = on
+      # Backtracking queries will look back for this amount of time. It should
+      # not be larger than the akka.projection.dynamodb.offset-store.time-window.
+      window = 2 minutes
+      # Backtracking queries read events up to this duration from the current database time.
```
Comment (pvlugter) [inline core/src/main/resources/reference.conf]:
```suggestion
      # Backtracking queries read events up to this duration from the current time.
```

**Review by pvlugter — APPROVED** [review]
LGTM. Also read parts of this as a diff with the r2dbc version, to see what's been adapted.

**Pull request merged**